### ISSUE TITLE
The Unitary matrix in Task2 is NOT available in dataset

### ISSUE THREAD

**Issue by BayanIssa** [Lucky-Lance/QAS-Bench#1]
Thank you for this interesting work. It was difficult for me to find an open source code with its relevant paper in QAS field, but with your nice repository it become easier to work in the field. 
However, I find an issue that while the codes of Genetic and other baselines use the unitary matrix as necessary input with path named "matrix_path", the dataset that provided in the link in description (https://drive.google.com/drive/folders/1H4McsQkDzruVmYyhd2PaEep35C49LriB?usp=sharing) do Not include the unitary matrix. It is only includes a pairs of training and validation set. 
Please provide the unitary matrix that correspond to Task2.

**Comment by Lucky-Lance** [comment]
Hi, thanks for the interest in our work, and sorry for the confusion.

This paper is a few years old, so some implementation details aren't fresh in my memory. But I checked the released files again, and the Task 2 unitary is in the dataset — it just isn't stored as a separate file. Each `*_train.pkl` / `*_valid.pkl` is a Python list where

- `[0]` is the target unitary, a `torch.complex128` tensor of shape \(2^n \times 2^n\);
- `[1:]` are the `{'input', 'output'}` state pairs, with `output = U @ input` (Eq. 8 in the paper).

The train and valid pickles of a given task hold the same matrix.

That's why `--matrix_path` points at the train pickle. The Task 2 scripts in this repo do

```python
U = pickle.load(open(args.matrix_path, "rb"))[0]
```

to build the search objective, and evaluate the metric \(f\) on `valid_data[1:]`. So the README command should work as-is:

```bash
python task2/Genetic/Genetic.py \
  --qubit_num 3 --layer_num 10 \
  --matrix_path data/task2/3bit_arbit/0/0_train.pkl \
  --valid_path data/task2/3bit_arbit/0/0_valid.pkl \
  --output_path result.pkl
```

Note that these paths assume the downloaded dataset is placed under `data/`, so that the folders become `data/task2/2bit_arbit`, `data/task2/3bit_arbit`, and so on.

If you want the matrix on its own:

```python
import pickle
data = pickle.load(open("0_train.pkl", "rb"))
U, pairs = data[0], data[1:]   # U: 8x8 complex128 tensor for 3 qubits
```

The name `matrix_path` is misleading here. In Task 1 the pickle does contain only the matrix, which is where the name comes from, but in Task 2 the matrix sits at index 0 of the list.

**Comment by BayanIssa** [comment]
Thank you very much for the fast response. The instruction above is working very well.
